src/librustc_data_structures/obligation_forest/mod.rs

```diff
@@ -162,38 +206,119 @@ pub struct ObligationForest<O: ForestObligation> {
     ///
     /// [details]: https://github.com/rust-lang/rust/pull/53255#issuecomment-421184780
     error_cache: FxHashMap<ObligationTreeId, FxHashSet<O::CacheKey>>,
+
+    /// Stores which nodes would be unblocked once `O::Variable` is unified
+    stalled_on: FxHashMap<O::Variable, Vec<NodeIndex>>,
+    /// Stores the node indices that are unblocked and should be processed at the next opportunity
+    unblocked: BinaryHeap<Unblocked>,
+    /// Stores nodes which should be processed on the next iteration since the variables they are
+    /// actually blocked on are unknown
+    stalled_on_unknown: Vec<NodeIndex>,
+    /// The offset that this `ObligationForest` has registered. Should be de-registered before
+    /// dropping this forest.
+    watcher_offset: Option<O::WatcherOffset>,
+    /// Reusable vector for storing unblocked nodes whose watch should be removed
+    temp_unblocked_nodes: Vec<O::Variable>,
+}
+
+/// Helper struct for use with `BinaryHeap` to process nodes in the order that they were added to
+/// the forest
+struct Unblocked {
+    index: NodeIndex,
+    order: u32,
+}
+
+impl PartialEq for Unblocked {
+    fn eq(&self, other: &Self) -> bool {
+        self.order == other.order
+    }
+}
+impl Eq for Unblocked {}
+impl PartialOrd for Unblocked {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        other.order.partial_cmp(&self.order)
+    }
+}
+impl Ord for Unblocked {
+    fn cmp(&self, other: &Self) -> Ordering {
+        other.order.cmp(&self.order)
+    }
 }
 
 #[derive(Debug)]
-struct Node<O> {
+struct Node<O: ForestObligation> {
     obligation: O,
     state: Cell<NodeState>,
 
+    /// A predicate (and its key) can change during processing. If it does we need to register the
+    /// old predicate so that we can remove or mark it as done if this node errors or is done.
+    alternative_predicates: Vec<O::CacheKey>,
+
     /// Obligations that depend on this obligation for their completion. They
     /// must all be in a non-pending state.
-    dependents: Vec<usize>,
+    dependents: Vec<NodeIndex>,
+    /// Obligations that this obligation depends on for their completion.
+    reverse_dependents: Vec<NodeIndex>,
 
     /// If true, `dependents[0]` points to a "parent" node, which requires
     /// special treatment upon error but is otherwise treated the same.
     /// (It would be more idiomatic to store the parent node in a separate
-    /// `Option<usize>` field, but that slows down the common case of
+    /// `Option<NodeIndex>` field, but that slows down the common case of
     /// iterating over the parent and other descendants together.)
     has_parent: bool,
 
     /// Identifier of the obligation tree to which this node belongs.
     obligation_tree_id: ObligationTreeId,
+    node_number: u32,
```

Comment:
This field could still use a comment (also, why a u32 and not a usize)?

Reply:
i.e., what is a "node number"? Just an arbitrary index that we assign to each node as it is created, I think?

Reply:
I commented the counter in `ObligationForest` itself but I can copy it here. I just used u32 since it doesn't have any connection to memory size. This `u32/i32` seems like the sensible default?

Reply:
OK, referencing the other field would be fine in the comment. I don't really care about u32/i32, just curious if there was a specific reason.